.github/workflows/code.yml

```
@@ -0,0 +1,191 @@
+        with:
+          fetch-depth: 0
+
+      - name: Cache Docker layers
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+
+      - name: Log in to GitHub Docker Registry
+        if: github.event_name != 'pull_request'
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ghcr.io/${{ github.repository }}
+          tags: |
+            type=ref,event=branch
+            type=ref,event=tag
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build developer image for testing
+        uses: docker/build-push-action@v3
+        with:
+          tags: build:latest
+          context: .
+          target: build
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache
+          load: true
+
+      - name: Run tests in the container locked with requirements_dev.txt
+        run: |
+          docker run --name test build bash /project/tests/container_tests.sh
+          docker cp test:/project/dist .
+          docker cp test:/project/requirements.txt .
+          docker cp test:/project/requirements_dev.txt .
+
+      - name: Build runtime image
+        uses: docker/build-push-action@v3
+        with:
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
+          context: .
+          labels: ${{ steps.meta.outputs.labels }}
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache
+
+      - name: Upload build files
+        uses: actions/upload-artifact@v3
+        with:
+          name: build_files
+          path: |
+            dist
+            requirements.txt
+            requirements_dev.txt
+
+  sdist:
+    needs: container
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.10"
+
+      - uses: actions/download-artifact@v3
+
+      - name: Install sdist in a venv and check cli works
+        # ${GITHUB_REPOSITORY##*/} is the repo name without org
+        # Replace this with the cli command if different to the repo name
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install build_files/dist/*.gz
+          ${GITHUB_REPOSITORY##*/} --version
+
+  release:
+    # upload to PyPI and make a release on every tag
+    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
+    needs: [lint, container]
```

I think we should still release if linting fails

```suggestion
    needs: container
```

---

it should be easy to fix linting and I'm probably about to do it - so why have a redundant release?

---

Cos if the linting fails on a tag because of a badly timed flake8 update, I think we should still publish the wheel

---

OK